Concurrent use of `scalaJSLinker` in sbt plugin

The Scala.js sbt plugin does not ensure that a project's linker instance is not used concurrently. A typical scenario where this arises is the following (discovered thanks to @reid-spencer):

```scala
val client = project.enablePlugins(ScalaJSPlugin, ScalaJSPlay)
val server = project.enablePlugins(PlayScalaJS).settings(scalaJSProjects += client)
val root = project.aggregate(client, server)
```
Then invoke:
```sh
sbt test
```

sbt now concurrently invokes the `client/test:loadedJSEnv` and the `client/test:fastOptJS` task. The former in order to detect the necessary tests for the `client` project, the latter to satisfy `server`'s dependency<sup id="a1">[1](#f1)</sup>.

These tasks in turn invoke `linkUnit`, `link` respectively on `client/test:scalaJSLinker` concurrently. This is not an allowed use of the linker. "Fortunately" an assert in `GenIncOptimizer` caught this:

```
java.lang.AssertionError: assertion failed
        at scala.Predef$.assert(Predef.scala:165)
        at org.scalajs.core.tools.linker.frontend.optimizer.GenIncOptimizer.withLogger(GenIncOptimizer.scala:97)
        at org.scalajs.core.tools.linker.frontend.optimizer.GenIncOptimizer.update(GenIncOptimizer.scala:107)
        at org.scalajs.core.tools.linker.frontend.LinkerFrontend$$anonfun$5.apply(LinkerFrontend.scala:71)
        at org.scalajs.core.tools.linker.frontend.LinkerFrontend$$anonfun$5.apply(LinkerFrontend.scala:71)
        at org.scalajs.core.tools.logging.Logger$class.time(Logger.scala:28)
        at org.scalajs.sbtplugin.Implicits$SbtLoggerWrapper.time(Implicits.scala:9)
```

# Workarounds

- Disable Rhino entirely (`scalaJSUseRhino in Global` := false). This will cause `loadedJSEnv` to depend on the `fastOptJS` task as well and sbt will properly linearize it.
- Disable sbt's parallel execution (`parallelExecution := false`). Exact required configuration and whether this works at all is untested.

# Possible fixes
## Teach sbt to not run the tasks concurrently
This is probably the optimal solution if feasible. It might be doable by using tags. However, [sbt's docs](http://www.scala-sbt.org/0.13/docs/Parallel-Execution.html) present tags more like a "global" thing, so it is unclear whether having a tag per configuration is feasible / reasonable and whether the sbt's scheduler can cope with so many tags.

## Synchronize usage of the linker
Dirty ad-hoc solution that completely ignores sbt's scheduler. Probably "good enough", but might expose speed degradations due to non-optimal scheduling (tasks blocking on other tasks while occupying execution slots).

---
<sup id="f1">1</sup>sbt-play-scalajs makes `server/test:resourceGenerators` depend on `client/test:fastOptJS`. [⬆](#a1)

--- a/sbt-plugin-test/build.sbt
+++ b/sbt-plugin-test/build.sbt
@@ -27,7 +27,11 @@ lazy val noDOM = project.settings(baseSettings: _*).
     scalaJSOutputWrapper := (
         "// Scala.js - noDOM sbt test\n//\n// Compiled with Scala.js\n",
         "// End of Scala.js generated script")
-  )
+  ).
+  /* This hopefully exposes concurrent uses of the linker. If it fails/gets
+   * flaky, there is a bug somewhere - #2202
+   */
+  settings(inConfig(Compile)(run <<= run.dependsOn(fastOptJS, loadedJSEnv)): _*)
 
 lazy val withDOM = project.settings(baseSettings: _*).
   enablePlugins(ScalaJSPlugin).

--- a/sbt-plugin/src/main/scala/scala/scalajs/sbtplugin/ScalaJSPluginInternal.scala
+++ b/sbt-plugin/src/main/scala/scala/scalajs/sbtplugin/ScalaJSPluginInternal.scala
@@ -59,6 +59,15 @@ object ScalaJSPluginInternal {
   val scalaJSLinker = SettingKey[ClearableLinker]("scalaJSLinker",
       "Scala.js internal: Setting to persist a linker", KeyRanks.Invisible)
 
+  /** A tag to indicate that a task is using the value of [[scalaJSLinker]]
+   *
+   *  This setting's value should always be retrieved from the same scope than
+   *  [[scalaJSLinker]] was retrieved from.
+   */
+  val usesScalaJSLinkerTag = SettingKey[Tags.Tag]("usesScalaJSLinkerTag",
+      "Scala.js internal: Tag to indicate that a task uses the link or " +
+      "linkUnit method of the value of scalaJSLinker", KeyRanks.Invisible)
+
   val scalaJSIRCacheHolder = SettingKey[globalIRCache.Cache]("scalaJSIRCacheHolder",
       "Scala.js internal: Setting to persist a cache. Do NOT use this directly. " +
       "Use scalaJSIRCache instead.", KeyRanks.Invisible)
@@ -189,7 +198,23 @@ object ScalaJSPluginInternal {
         new ClearableLinker(newLinker, opts.batchMode)
       },
 
-      key := {
+      usesScalaJSLinkerTag in key := {
+        val projectPart = thisProject.value.id
+        val configPart = configuration.value.name
+
+        val stagePart = stage match {
+          case Stage.FastOpt => "fastopt"
+          case Stage.FullOpt => "fullopt"
+        }
+
+        Tags.Tag(s"uses-scalajs-linker-$projectPart-$configPart-$stagePart")
+      },
+
+      // Prevent this linker from being used concurrently
+      concurrentRestrictions in Global +=
+        Tags.limit((usesScalaJSLinkerTag in key).value, 1),
+
+      key <<= Def.taskDyn {
         val s = (streams in key).value
         val log = s.log
         val irInfo = (scalaJSIR in key).value
@@ -197,27 +222,29 @@ object ScalaJSPluginInternal {
         val ir = irInfo.data
         val output = (artifactPath in key).value
 
-        FileFunction.cached(s.cacheDirectory, FilesInfo.lastModified,
-            FilesInfo.exists) { _ => // We don't need the files
+        Def.task {
+          FileFunction.cached(s.cacheDirectory, FilesInfo.lastModified,
+              FilesInfo.exists) { _ => // We don't need the files
 
-          val stageName = stage match {
-            case Stage.FastOpt => "Fast"
-            case Stage.FullOpt => "Full"
-          }
+            val stageName = stage match {
+              case Stage.FastOpt => "Fast"
+              case Stage.FullOpt => "Full"
+            }
 
-          log.info(s"$stageName optimizing $output")
+            log.info(s"$stageName optimizing $output")
 
-          IO.createDirectory(output.getParentFile)
+            IO.createDirectory(output.getParentFile)
 
-          val linker = (scalaJSLinker in key).value
-          linker.link(ir, AtomicWritableFileVirtualJSFile(output), log)
+            val linker = (scalaJSLinker in key).value
+            linker.link(ir, AtomicWritableFileVirtualJSFile(output), log)
 
-          logIRCacheStats(log)
+            logIRCacheStats(log)
 
-          Set(output)
-        } (realFiles.toSet)
+            Set(output)
+          } (realFiles.toSet)
 
-        Attributed.blank(output)
+          Attributed.blank(output)
+        } tag((usesScalaJSLinkerTag in key).value)
       },
 
       key <<= key.dependsOn(packageJSDependencies, packageScalaJSLauncher),
@@ -332,7 +359,8 @@ object ScalaJSPluginInternal {
       scalajspSettings ++
       stageKeys.flatMap((scalaJSStageSettings _).tupled) ++
       dispatchTaskKeySettings(scalaJSLinkedFile) ++
-      dispatchSettingKeySettings(scalaJSLinker)
+      dispatchSettingKeySettings(scalaJSLinker) ++
+      dispatchSettingKeySettings(usesScalaJSLinkerTag)
   ) ++ Seq(
       /* Note: This cache only gets freed by its finalizer. Otherwise we'd need
        * to intercept reloads in sbt (see #2171).
@@ -558,7 +586,7 @@ object ScalaJSPluginInternal {
 
               log.debug("Loading JSEnv with LinkingUnit")
               env.loadLibs(libs).loadLinkingUnit(unit)
-            }
+            } tag(usesScalaJSLinkerTag.value)
           case env =>
             Def.task {
               val file = scalaJSLinkedFile.value

